Fixed #30803 -- Allowed comma separators for milliseconds in django.utils.dateparse functions.

farhaanbukhsh opened the pull request:
This is a continuation to complete https://github.com/django/django/pull/11818

charettes commented on docs/ref/utils.txt:
```diff
@@ -145,9 +145,10 @@ The functions defined in this module share the following properties:
+    Expects data in the format ``"DD HH:MM:SS.uuuuuu"`` or ``"DD HH:MM:SS,uuuuuu"``
+    or as specified by ISO 8601 (e.g. ``P4DT1H15M20S`` which is equivalent to
+    ``4 1:15:20``) or PostgreSQL's day-time interval format
+    (e.g. ``3 days 04:05:06``).
```
There should likely be a `::versionadded 3.1` here for the new accepted format. And maybe a mention in the release notes?

farhaanbukhsh replied:
how do I do a '::versionadded 3.1' ? I will find out how to add release notes to it.

charettes replied:
https://docs.djangoproject.com/en/2.2/internals/contributing/writing-documentation/#documenting-new-features

MoonLover commented:
+1

felixxm commented on docs/ref/utils.txt:
```diff
@@ -145,9 +145,10 @@ The functions defined in this module share the following properties:
-    Expects data in the format ``"DD HH:MM:SS.uuuuuu"`` or as specified by ISO
-    8601 (e.g. ``P4DT1H15M20S`` which is equivalent to ``4 1:15:20``) or
-    PostgreSQL's day-time interval format (e.g. ``3 days 04:05:06``).
+    Expects data in the format ``"DD HH:MM:SS.uuuuuu"`` or ``"DD HH:MM:SS,uuuuuu"``
```
``` ``"DD HH:MM:SS.uuuuuu"`` or ``"DD HH:MM:SS,uuuuuu"`` or ``` ->
``` ``"DD HH:MM:SS.uuuuuu"``, ``"DD HH:MM:SS,uuuuuu"``, or ```

Please wrap docs at 79 chars.

felixxm commented:
@farhaanbukhsh I squashed commits, pushed minor edits, updated docs and added two tests.

felixxm commented on tests/utils_tests/test_dateparse.py:
```diff
@@ -131,10 +135,12 @@ def test_iso_8601(self):
             ('PT5M', timedelta(minutes=5)),
             ('PT5S', timedelta(seconds=5)),
             ('PT0.000005S', timedelta(microseconds=5)),
+            ('PT0,000005S', timedelta(microseconds=5)),
```
I added these test because we should support comma separators in this format.

farhaanbukhsh commented:
> @farhaanbukhsh I squashed commits, pushed minor edits, updated docs and added two tests.

Thanks a lot, I learnt a lot of things working on this PR. 😄

farhaanbukhsh commented on django/utils/dateparse.py:
```diff
@@ -137,11 +137,11 @@ def parse_duration(value):
-        kw = {k: float(v) for k, v in kw.items() if v is not None}
+        kw = {k: float(v.replace(',', '.')) for k, v in kw.items() if v is not None}
+        days = datetime.timedelta(kw.pop('days', .0) or .0)
         return days + sign * datetime.timedelta(**kw)
```
Hey @felixxm, I have a doubt rather a clarification, I found some variables a little less informative in terms of what they are doing, like the  variable `kw` which I believe is coming from `kwargs` but wouldn't it make more sense if we name it something like `time_map` or `date_info`?

Same in dictionary comprehension that we are using.

Is there are specific reason we kept it like that?

Just curious about it :) Thanks in advance

felixxm replied:
They aren't misleading IMO and we try to avoid trivial changes.

felixxm commented:
Merged in 42b23d1e79b21ddd8abd7216a6771f105dd7062c and 1f817daa20110987bf155fc6c942a8a14720a66b.
